raft_topology, tablets: Drain tablets in parallel with other topology operations

tgrabiec opened the pull request:
Allows bootstrap to execute while tablets are being drained on decommission. Allows multiple decommission/removenode to happen in parallel.

Flow of decommission/removenode request:
  1) pending, has tablet replicas on target node
  2) not pending, paused (node::is_drained() == true)
  3) pending, no tablet replicas on target node
  4) executing, node is in transition
  5) done

Request is unpaused when topology state machine is idle, which makes the request pending and ready to be scheduled. Tablet load balancing will yield the state machine as soon as some request can be unpaused.

TODO:
 1) Failure propagation
 2) API for canceling requests
 3) Unit tests

Fixes #21452

cezarmoise commented:
Tested decommissioning 2 nodes x 3 racks
https://argus.scylladb.com/tests/scylla-cluster-tests/174be212-b9ac-4fc7-ae20-8b659bfb53ec
![image](https://github.com/user-attachments/assets/e0b985b5-e6bb-42d5-a441-f406fbd9c97e)

tgrabiec commented on service/topology_state_machine.hh:
```diff
@@ -110,7 +112,7 @@ struct topology {
         join_group0,
         commit_cdc_generation,
-        tablet_draining,
+        tablet_draining, // deprecated, not set after feature_service::parallel_decommission is enabled.
```
You mean a transition_state? No, because the point of the change is to release topology state machine to other things, like bootstrap, while we drain.

mykaul commented:
@gleb-cloudius - can you help and take this over and rebase over your changes in this area?

gleb-cloudius commented:
> @gleb-cloudius - can you help and take this over and rebase over your changes in this area?

There should not be any conflicts. I do not think there were changes in this area.

bhalevy commented:
@tgrabiec claimed that some business logic changes would be required after the rebase.
Please elaborate.

cezarmoise commented:
> @tgrabiec claimed that some business logic changes would be required after the rebase. Please elaborate.

I tried to cherry-pick this commit into the branch with out of space changes, and indeed only this conflict had to be resolved
> Indeed, all I see is this:
> 
> ```
> <<<<<<< parallel-decomm
>     gms::feature parallel_decommission { *this, "PARALLEL_DECOMMISSION"sv };
> =======
>     gms::feature load_and_stream_abort_rpc_message { *this, "LOAD_AND_STREAM_ABORT_RPC_MESSAGE"sv };
>     gms::feature topology_global_request_queue { *this, "TOPOLOGY_GLOBAL_REQUEST_QUEUE"sv };
>     gms::feature lwt_with_tablets { *this, "LWT_WITH_TABLETS"sv };
>     gms::feature repair_msg_split { *this, "REPAIR_MSG_SPLIT"sv };
> >>>>>>> master
> ```

but it failed to compile with
```
16:49:50  FAILED: build/release/service/topology_coordinator.o 
16:49:50  clang++ -MD -MT build/release/service/topology_coordinator.o -MF build/release/service/topology_coordinator.o.d -std=gnu++23 -I/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla/seastar/include -I/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla/build/release/seastar/gen/include -Werror=unused-result -DSEASTAR_API_LEVEL=7 -DSEASTAR_SSTRING -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=19 -DSEASTAR_LOGGER_TYPE_STDOUT -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_THREAD_NO_LIB -DBOOST_THREAD_DYN_LINK -DFMT_SHARED -I/usr/include/p11-kit-1 -g -ffile-prefix-map=/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -march=westmere -Wno-backend-plugin -flto=thin -ffat-lto-objects -fprofile-use=/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla/build/profile.profdata -mllvm -pgso=false -mllvm -enable-value-profiling=false -ffunction-sections -fdata-sections  -O3 -mllvm -inline-threshold=2500 -fno-slp-vectorize -DSCYLLA_BUILD_MODE=release -g -gz -Xclang -fexperimental-assignment-tracking=disabled -iquote. -iquote build/release/gen -std=gnu++23 -g -ffile-prefix-map=/jenkins/workspace/scylla-master/byo/byo_build_tests_dtest/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -march=westmere -I/usr/include/p11-kit-1 -DBOOST_ALL_DYN_LINK    -fvisibility=hidden -Ikmipc/kmipc-2.1.0t-rhel84_64/include -DHAVE_KMIP -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN  -c -o build/release/service/topology_coordinator.o service/topology_coordinator.cc
16:49:50 
 service/topology_coordinator.cc:2068:58: error: no viable conversion from 'typename std::remove_reference<vector<canonical_mutation> &>::type' (aka 'std::vector<canonical_mutation>') to 'utils::chunked_vector<canonical_mutation>'
16:49:50   2068 |         co_await update_topology_state(std::move(guard), std::move(muts), "unpause requests");
16:49:50        |                                                          ^~~~~~~~~~~~~~~
16:49:50  ./utils/chunked_vector.hh:102:5: note: candidate constructor not viable: no known conversion from 'typename std::remove_reference<vector<canonical_mutation> &>::type' (aka 'std::vector<canonical_mutation>') to 'const chunked_vector<canonical_mutation> &' for 1st argument
16:49:50    102 |     chunked_vector(const chunked_vector& x);
16:49:50        |     ^              ~~~~~~~~~~~~~~~~~~~~~~~
16:49:50  ./utils/chunked_vector.hh:104:5: note: candidate constructor not viable: no known conversion from 'typename std::remove_reference<vector<canonical_mutation> &>::type' (aka 'std::vector<canonical_mutation>') to 'chunked_vector<canonical_mutation> &&' for 1st argument
16:49:50    104 |     chunked_vector(chunked_vector&& x) noexcept;
16:49:50        |     ^              ~~~~~~~~~~~~~~~~~~
16:49:50  ./utils/chunked_vector.hh:112:5: note: candidate constructor not viable: no known conversion from 'typename std::remove_reference<vector<canonical_mutation> &>::type' (aka 'std::vector<canonical_mutation>') to 'std::initializer_list<canonical_mutation>' for 1st argument
16:49:50    112 |     chunked_vector(std::initializer_list<T> x);
16:49:50        |     ^              ~~~~~~~~~~~~~~~~~~~~~~~~~~
16:49:50  ./utils/chunked_vector.hh:113:14: note: explicit constructor is not a candidate
16:49:50    113 |     explicit chunked_vector(size_t n, const T& value = T());
16:49:50        |              ^
16:49:50  service/topology_coordinator.cc:374:77: note: passing argument to parameter 'updates' here
16:49:50    374 |             group0_guard guard, utils::chunked_vector<canonical_mutation>&& updates, const sstring& reason) {
16:49:50        |                                                                             ^
16:49:50  1 error generated.
```

tgrabiec commented:
> > @tgrabiec claimed that some business logic changes would be required after the rebase. Please elaborate.
> 
> I tried to cherry-pick this commit into the branch with out of space changes, and indeed only this conflict had to be resolved
> 
> > Indeed, all I see is this:
> > ```
> > <<<<<<< parallel-decomm
> >     gms::feature parallel_decommission { *this, "PARALLEL_DECOMMISSION"sv };
> > =======
> >     gms::feature load_and_stream_abort_rpc_message { *this, "LOAD_AND_STREAM_ABORT_RPC_MESSAGE"sv };
> >     gms::feature topology_global_request_queue { *this, "TOPOLOGY_GLOBAL_REQUEST_QUEUE"sv };
> >     gms::feature lwt_with_tablets { *this, "LWT_WITH_TABLETS"sv };
> >     gms::feature repair_msg_split { *this, "REPAIR_MSG_SPLIT"sv };
> > >>>>>>> master
> > ```
> 
> but it failed to compile with

Try now.

tgrabiec commented:
> @tgrabiec claimed that some business logic changes would be required after the rebase. Please elaborate.

I suspected there could be logical conflicts with "topology coordinator: Implement global topology request queue", but after looking at it, looks like there are no conflicts.

dani-tweig commented:
@tgrabiec , @bhalevy , what are the remaining steps here?
Will it make the branch date of 2025.4 (Sep 30)?

tgrabiec commented:
In V2:
  - fix removenode hanging
  - make failure to drain fail the topology request
  - support for cancellation via task API 
  - added unit tests

Deexie commented on service/topology_state_machine.cc:
```diff
@@ -261,6 +261,49 @@ future<sstring> topology_state_machine::wait_for_request_completion(db::system_k
+    tsmlogger.info("Will cancel {} request {} for node {}: {}", req, request_id, node, reason);
+
+    topology_mutation_builder builder(guard.write_timestamp());
+    auto node_builder = builder.with_node(node)
```
`auto&`

Deexie commented on service/topology_state_machine.cc:
```diff
@@ -261,6 +261,49 @@ future<sstring> topology_state_machine::wait_for_request_completion(db::system_k
+    tsmlogger.info("Will cancel {} request {} for node {}: {}", req, request_id, node, reason);
+
+    topology_mutation_builder builder(guard.write_timestamp());
+    auto node_builder = builder.with_node(node)
```
`auto&`?

tgrabiec replied:
Right

tgrabiec commented on node_ops/task_manager_module.cc:
```diff
@@ -106,8 +106,8 @@ future<std::optional<tasks::virtual_task_hint>> node_ops_virtual_task::contains(
 
-    auto entry = co_await _ss._sys_ks.local().get_topology_request_entry(task_id.uuid(), false);
-    co_return bool(entry.id) && std::holds_alternative<service::topology_request>(entry.request_type) ? empty_hint : std::nullopt;
+    auto entry = co_await _ss._sys_ks.local().get_topology_request_entry(task_id.uuid());
```
Right, bad patch splitting.

tgrabiec commented:
> > > @tgrabiec how about documentation? Should we update e.g. https://docs.scylladb.com/manual/stable/operating-scylla/nodetool-commands/decommission.html and https://docs.scylladb.com/manual/stable/operating-scylla/nodetool-commands/removenode.html ?
> > 
> > 
> > We could, or we could treat it as implementation detail. They already can be invoked in parallel, we just serialize them now.
> 
> I would vote for updating, it is a big improvement and we should propagate it in documentation and in release notes

Ok, I'll documentation in the next version.

tgrabiec commented:
> #### 🔴 CI State: FAILURE
> Mode	Stage	Status	Node
> [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22266/)	✅	sif
> dev	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22266/)	✅	spider5.cloudius-systems.com
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22266/)	❌	spider5.cloudius-systems.com
> release	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22266/)	✅	spider15.cloudius-systems.com
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22266/)	❌	spider15.cloudius-systems.com
> debug	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22266/)	✅	spider5.cloudius-systems.com
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22266/)	❌	spider5.cloudius-systems.com
> dtest	[dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/21015/)	✅	
> [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/9470/)	❌	
> [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/12788/)	✅	
> #### Failed Tests (15/78773):
> * [test_mutation_schema_change.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/22266/testReport/junit/cluster/test_mutation_schema_change/debug___test_py___debug___test_mutation_schema_change_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mutation_schema_change.debug.1)

https://github.com/scylladb/scylladb/issues/27422

cezarmoise commented on service/tablet_allocator.cc:
```diff
@@ -3129,15 +3140,15 @@ class load_balancer {
         if (!min_load_node) {
-            if (!nodes_to_drain.empty()) {
-                throw std::runtime_error(format("There are nodes with tablets to drain but no candidate nodes in DC {}."
-                                                " Consider adding new nodes or reducing replication factor.", dc));
```
Changing this error message is what causes this test to fail

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/9468/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/FullDtest___full_split000___test_decommission_with_multi_rack/

tgrabiec commented:
> ```
> task_id                              type         kind    scope   state  sequence_number keyspace table entity shard start_time           end_time            
> d9f47f40-d130-11f0-9b50-ca49e0f227b0 bootstrap    cluster cluster done   0                                     0     2025-12-04T16:47:00Z 2025-12-04T16:47:01Z
> 0af42e1a-d131-11f0-a90f-c91cb7ec6e34 decommission cluster cluster failed 0                                     0                          2025-12-04T16:48:22Z
> ```
> 
> and start time is missing, compared to a successful decommission
> 
> ```
> task_id                              type         kind    scope   state sequence_number keyspace table entity shard start_time           end_time            
> 29dd0a6c-d1cd-11f0-be18-37d24e50e5bb decommission cluster cluster done  0                                     0     2025-12-05T11:25:57Z 2025-12-05T11:26:04Z
> 17e7bf50-d1cd-11f0-88b9-1242e446f5a6 bootstrap    cluster cluster done  0                                     0     2025-12-05T11:25:25Z 2025-12-05T11:25:26Z
> ```

That's strange. TTL is set to 1 month. I checked the code, and we never delete "start_time" explicitly. My tests don't reproduce the issue, "start_time" is there.

Is it reproducible every time?

tgrabiec commented:
> > ```
> > task_id                              type         kind    scope   state  sequence_number keyspace table entity shard start_time           end_time            
> > d9f47f40-d130-11f0-9b50-ca49e0f227b0 bootstrap    cluster cluster done   0                                     0     2025-12-04T16:47:00Z 2025-12-04T16:47:01Z
> > 0af42e1a-d131-11f0-a90f-c91cb7ec6e34 decommission cluster cluster failed 0                                     0                          2025-12-04T16:48:22Z
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and start time is missing, compared to a successful decommission
> > ```
> > task_id                              type         kind    scope   state sequence_number keyspace table entity shard start_time           end_time            
> > 29dd0a6c-d1cd-11f0-be18-37d24e50e5bb decommission cluster cluster done  0                                     0     2025-12-05T11:25:57Z 2025-12-05T11:26:04Z
> > 17e7bf50-d1cd-11f0-88b9-1242e446f5a6 bootstrap    cluster cluster done  0                                     0     2025-12-05T11:25:25Z 2025-12-05T11:25:26Z
> > ```
> 
> That's strange. TTL is set to 1 month. I checked the code, and we never delete "start_time" explicitly. My tests don't reproduce the issue, "start_time" is there.
> 
> Is it reproducible every time?

I managed to reproduce it without abort being involved. It probably means that you canceled the task after it was queued but before it started executing. topology coordinator sets "start_time" when the request is picked up for execution from the pending state.

bhalevy commented:
> I can see that different requests have a different idea when "start_time" should be set. Some set it at queue time. We should unify this.

Makes sense.
Can you please open an issue about that.

Cc @denesb @Deexie

tgrabiec commented:
In V6:
 - fix failure in cluster/test_topology_ops::test_topology_ops, by ensuring no co-locating migrations are emitted on excluded && drained nodes (being removed)

New commits:
 - test: tablets: Check that there are no migrations scheduled on drained nodes
 - test: lib: topology_builder: Introduce set_drained()
 - tablet_allocator: Skip co-location on drained nodes

Deexie commented on db/virtual_tables.cc:
```diff
@@ -68,6 +68,8 @@ class cluster_status_table : public memtable_filling_virtual_table {
             .with_column("peer", inet_addr_type, column_kind::partition_key)
             .with_column("dc", utf8_type)
             .with_column("up", boolean_type)
+            .with_column("drained", boolean_type)
```
I assume we don't care about the fact that during upgrade new nodes show more columns than the old ones, do we?

Deexie commented on service/tablet_allocator.cc:
```diff
@@ -2832,6 +2837,9 @@ class load_balancer {
                     for (auto&& [skip_info, tablets] : *skip) {
                         process_skip_info(tablets, skip_info);
                     }
+                    if (!plan.drain_failures().empty()) {
```
Why not check it at the beginning of the loop?

Deexie commented on service/tablet_allocator.hh:
```diff
@@ -141,12 +154,14 @@ private:
     const migrations_vector& migrations() const { return _migrations; }
-    bool empty() const { return _migrations.empty() && !_resize_plan.size() && !_repair_plan.size();}
+    const std::vector<drain_failure>& drain_failures() const { return _drain_failures; }
+    bool empty() const { return _migrations.empty() && !_resize_plan.size() && !_repair_plan.size() && _drain_failures.empty(); }
```
How about `size()`?

tgrabiec replied:
Right. I think we should drop size(), as it's meaningless. It used to count migrations, but now it's not, and it's used for emptiness, but for that empty() is enough.

Deexie commented on service/topology_coordinator.cc:
```diff
@@ -1339,9 +1346,30 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+            auto drain_failures = plan.drain_failures();
+
+            // Need to cancel decommission requests if nodes got excluded to avoid unintended rebuilds.
+            for (auto [node, req] : _topo_sm._topology.paused_requests) {
```
Why don't we do it in `make_plan` instead?

Deexie commented on service/topology_coordinator.cc:
```diff
@@ -1339,9 +1346,30 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+            // Need to cancel decommission requests if nodes got excluded to avoid unintended rebuilds.
+            for (auto [node, req] : _topo_sm._topology.paused_requests) {
+                if (req == topology_request::leave && _topo_sm._topology.excluded_tablet_nodes.contains(node)) {
+                    drain_failures.push_back(drain_failure(locator::host_id(node.uuid()), "decommissioned node is excluded"));
```
We may generate_cancel_request_update more than once for the same node

Deexie commented on service/topology_coordinator.cc:
```diff
@@ -1339,9 +1346,30 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                    auto server_id = raft::server_id(drain_fail.node().uuid());
+                    _topo_sm.generate_cancel_request_update(out, _feature_service, guard, server_id, drain_fail.reason());
+                }
+            } else {
```
Since we omit all migrations if there are drain failures, `make_plan` can return early if there are any `drain_failures`

tgrabiec replied:
Yes, that already mostly happens. We break out of inter-node plan. Only intra-node plan may be generated.

We could optimize it, but it's a rare event.

Deexie commented on service/topology_coordinator.cc:
```diff
@@ -2912,6 +2975,7 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                 builder.set_fence_version(_topo_sm._topology.version) // fence requests in case the drain above failed
                        .set_transition_state(topology::transition_state::tablet_migration) // in case tablet drain failed we need to complete tablet transitions
                        .with_node(node.id)
+                       .set("drained", false)
```
`.set("paused", false)`

tgrabiec commented on service/topology_state_machine.cc:
```diff
@@ -289,6 +289,12 @@ void topology_state_machine::generate_cancel_request_update(utils::chunked_vecto
 
+    if (features.parallel_decommission) {
+        node_builder
+            .del("request_paused")
```
They're equivalent.

tgrabiec commented on node_ops/task_manager_module.cc:
```diff
@@ -147,9 +130,8 @@ future<> node_ops_virtual_task::abort(tasks::task_id id, tasks::virtual_task_hin
-    co_return std::ranges::to<std::vector<tasks::task_stats>>(co_await get_entries(sys_ks, topology, get_task_manager().get_user_task_ttl())
-            | std::views::transform([] (const auto& e) {
+    co_return std::ranges::to<std::vector<tasks::task_stats>>(co_await get_entries(sys_ks, get_task_manager().get_user_task_ttl())
+            | std::views::transform([)] (const auto& e) {
```
Yes, it's fixed in a later patch.
Slipped by because of over-trusting in copilot auto-completion.

tgrabiec replied:
fixed

patjed41 commented on service/topology_coordinator.cc:
```diff
@@ -2187,18 +2187,13 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
-            auto node_builder = builder.with_node(id).del("topology_request");
             auto done_msg = fmt::format("Canceled. Dead nodes: {}", dead_nodes);
-            rtbuilder.done(done_msg);
+            _topo_sm.generate_cancel_request_update(muts, _feature_service, guard, id, done_msg);
```
`_feature_service` is unused by `generate_cancel_request_update`.

It looks like we copy `done_msg` for no reason.

patjed41 commented on service/topology_state_machine.cc:
```diff
@@ -261,6 +261,55 @@ future<sstring> topology_state_machine::wait_for_request_completion(db::system_k
+    auto i = _topology.requests.find(node);
+    if (i == _topology.requests.end()) {
+        // Request is not pending, and not cancelable at this stage.
+        return;
```
The function also returns here when there is no pending/ongoing request at all, which may be surprising. That's what happens for normal nodes. Here is the reason: https://github.com/scylladb/scylladb/blob/73db5c94dee9945d252fba253f44c90c90a944ae/service/topology_state_machine.hh#L95
Perhaps we can extend this comment to make it clear. It currently looks as if the function returned in the if statement above when there is no pending/ongoing request.

patjed41 commented on locator/topology.hh:
```diff
@@ -139,6 +141,16 @@ public:
     }
 
+    // Drained nodes are about to be decommissioned and tablet scheduler
+    // moves tablet replicas away from them.
```
Not sure if I understand this correctly at this point of my review, but it looks like only a normal node that is about to be decommissioned can be drained. Shouldn't drain be a value of `node::state` then?

patjed41 commented on service/topology_coordinator.cc:
```diff
@@ -312,6 +315,10 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+            if (!topo.paused_requests.empty()) {
+                // Allow paused requests to finish.
+                return std::move(guard);
+            }
```
Why is it fine to not cancel requests that are not paused in this case? Why can't we cancel not paused requests while waiting for paused requests?

pehala commented on docs/operating-scylla/nodetool-commands/decommission.rst:
```diff
@@ -29,5 +29,13 @@ Before you run ``nodetool decommission``:
+this part is serialized with other vnode-based operations, including those from other decommission operations.
+
+Decommission which is still in the tablet draining phase can be canceled using Task Manager API.
+See :doc:`Task manager </operating-scylla/admin-tools/task-manager>`.
```
I would document the exact procedure here as well. The task manager doc is more an API overview and it is not clear what exactly you need to do to cancel it. Also by documenting it will became clear how involved the current process is.

I do not need this update to be in this PR, I think the current doc is sufficient, but we should improve the UX on using the cancellation and their respective docs as followups.

patjed41 commented on gms/feature_service.hh:
```diff
@@ -171,6 +171,7 @@ public:
     gms::feature topology_global_request_queue { *this, "TOPOLOGY_GLOBAL_REQUEST_QUEUE"sv };
     gms::feature lwt_with_tablets { *this, "LWT_WITH_TABLETS"sv };
     gms::feature repair_msg_split { *this, "REPAIR_MSG_SPLIT"sv };
+    gms::feature parallel_decommission { *this, "PARALLEL_DECOMMISSION"sv };
```
Maybe `parallel_decommission_removenode`?

tgrabiec replied:
parallel_tablet_draining seems better indeed.

patjed41 commented on service/topology_state_machine.cc:
```diff
@@ -316,6 +319,56 @@ void topology_state_machine::generate_cancel_request_update(utils::chunked_vecto
+        }
+
+        if (muts.empty()) {
+            // FIXME: Handle global requests.
```
Do we need it for something? If yes, let's open a follow-up issue.

patjed41 reviewed (commented):
Looks good overall, I have mostly minor comments and questions.

I still need to review the last few commits with tests and docs.

patjed41 commented on test/cluster/test_tablets_parallel_decommission.py:
```diff
@@ -0,0 +1,449 @@
+    the decommissions are indeed progressing in parallel and not sequentially.
+    """
+
+    cfg = {'enable_tablets': True}
```
It's unneeded now. We have `tablets_mode_for_new_keyspaces`, and it's set to true for cluster tests. The same goes for other test cases.

patjed41 commented on test/cluster/test_tablets_parallel_decommission.py:
```diff
@@ -0,0 +1,449 @@
+
+    coord_srv = servers[0]
+    servers_to_remove = [servers[2], servers[3]]
+    host_ids_to_remove = await asyncio.gather(*(manager.get_host_id(s.server_id) for s in servers_to_remove))
```
`gather_safely` is recommended

tgrabiec commented on service/topology_coordinator.cc:
```diff
@@ -1382,9 +1389,35 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+            }
+
+            // Need to cancel decommission requests if nodes got excluded to avoid unintended rebuilds.
+            for (auto [node, req] : _topo_sm._topology.paused_requests) {
```
> IIUC, we won't cancel a pending decommission request when the node becomes excluded. We will start draining and fail it here, or move directly to write_both_read_old and fail there. It seems strange. There is no reason to even start handling the request.

We could cancel the request earlier. But there are many entry points for marking the node as excluded, so cooperative abort is safer. I admit that it looks wasteful, but it doesn't seem to be a big problem, it will happen soon enough.

> I didn't know that making a decommissioning node excluded is even possible. It seems fishy to me because it can happen at any point of decommission and it's unclear how it will be handled in each possible case. Do we have it specified? Do we test it?

Nodes can be marked as excluded at any point using `nodetool exclude`, or `nodetool removenode` (we do it before request is executed). That's the first step of recovery from losing a node, which allows barriers to proceed. Tablet migrations would be stuck without it, as rollback requires a barrier too to restore tablet invariants.

We test it extensively for tablets in test_node_failure_during_tablet_migration.

patjed41 commented on test/pylib/tablets.py:
```diff
@@ -42,6 +43,18 @@ async def get_all_tablet_replicas(manager: ManagerClient, server: ServerInfo, ke
+
 async def get_tablet_replicas(manager: ManagerClient, server: ServerInfo, keyspace_name: str, table_name: str, token: int) -> list[tuple[HostID, int]]:
     """
     Gets tablet replicas of the tablet which owns a given token of a given table.
```
Some test cases may be missing:
- it's possible to bootstrap a node while a decommission request is paused,
- parallel removenode and decommission (not sure interesting).

But overall the tests seem very good and cover the problematic cases.

patjed41 commented on docs/operating-scylla/nodetool-commands/decommission.rst:
```diff
@@ -29,5 +29,13 @@ Before you run ``nodetool decommission``:
   request may fail.
   In such a case, ALTER the keyspace to reduce the RF before running ``nodetool decommission``.
 
+It's allowed to invoke ``nodetool decommission`` on multiple nodes in parallel. This will be faster than doing
```
We could mention that it's possible to bootstrap nodes during draining. Not sure what is a good place.

patjed41 commented:
@annastuchlik please review the last commit with doc changes

tgrabiec commented on service/topology_state_machine.hh:
```diff
@@ -111,7 +113,7 @@ struct topology {
         join_group0,
         commit_cdc_generation,
-        tablet_draining,
+        tablet_draining, // deprecated, not set after feature_service::parallel_decommission is enabled.
```
Adjusted diagrams. 

And added this:

```
+## Tablet draining
+
+Node requests of type `leave` (decommission) and `remove` start by setting `drained=true` for the
+target node and putting topology state machine in the `tablet_migration` transition. Tablet
+scheduler will work to migrate tablets away from those nodes.
+
+The node request is put in the paused state, and will be ignored by topology coordinator's scheduler.
+This allows other topology requests to preempt tablet balancing and execute.
+
+The node's state is still `normal` during tablet draining phase.
+
+When tablet scheduler is done with migrating all tablets away from a drained node,
+the associated request is unpaused, and can be picked by topology coordinator again
+for execution. This time it will enter the `write_both_read_old` transition and proceed
+with the vnode part.
+
+This allows multiple requests to drain tablets in parallel, and other requests to not be blocked
+by long `leave` and `remove` requests.
+

```

tgrabiec commented on service/storage_service.cc:
```diff
@@ -5025,6 +5025,10 @@ future<> storage_service::wait_for_topology_not_busy() {
     }
```
Not only, anything, including truncate or ALTER KEYSPACE. But bootstrap is interesting because of scale-out. I will improve the wording.

tgrabiec commented on test/pylib/tablets.py:
```diff
@@ -42,6 +43,18 @@ async def get_all_tablet_replicas(manager: ManagerClient, server: ServerInfo, ke
+
 async def get_tablet_replicas(manager: ManagerClient, server: ServerInfo, keyspace_name: str, table_name: str, token: int) -> list[tuple[HostID, int]]:
     """
     Gets tablet replicas of the tablet which owns a given token of a given table.
```
I think we should already have tests for decommission / removenode in a mixed cluster. It's not a new problem which this PR introduced. 

@pehala Do you have such tests?

tgrabiec commented on test/cluster/test_tablets_parallel_decommission.py:
```diff
@@ -0,0 +1,449 @@
+    logger.info("Stopping servers to be removed")
+
+    for srv in servers_to_remove:
+        await manager.server_stop_gracefully(srv.server_id)
```
Will do

tgrabiec commented on test/cluster/test_tablets_parallel_decommission.py:
```diff
@@ -0,0 +1,449 @@
+    ])
+
+    cql = manager.get_cql()
+    await cql.run_async("CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy',"
```
Right

tgrabiec commented on db/system_keyspace.cc:
```diff
@@ -260,6 +260,8 @@ schema_ptr system_keyspace::topology() {
             .with_column("release_version", utf8_type)
             .with_column("topology_request", utf8_type)
+            .with_column("request_paused", boolean_type)
+            .with_column("drained", boolean_type)
```
> why should a node request be ignored by the scheduler?

```
- `request_paused`     -  set to true when node request should be ignored by the scheduler because it's waiting for the node to be drained of tablet replicas
```

bhalevy commented:
> #### 🔴 CI State: FAILURE
> Mode	Stage	Status	Node
> [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/23023/)	✅	spider9.cloudius-systems.com
> dev	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/23023/)	❌	spider8.cloudius-systems.com
> release	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/23023/)	❌	spider1.cloudius-systems.com
> debug	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/23023/)	❌	spider15.cloudius-systems.com
> #### Build Details:
> * Duration: 2 hr 12 min
> 
> **Note:** To re-trigger CI for this PR, comment: `@scylladbbot trigger-ci`

@tgrabiec :
```
04:22:50  clang++ -MD -MT build/dev/test/boost/tablets_test.o -MF build/dev/test/boost/tablets_test.o.d -std=gnu++23 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=9 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=21 -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DFMT_SHARED -I/usr/include/p11-kit-1 -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -fextend-variable-liveness=none -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -DSCYLLA_ENABLE_PREEMPTION_SOURCE -fpch-validate-input-files-content -O2 -DSCYLLA_BUILD_MODE=dev -iquote. -iquote build/dev/gen -std=gnu++23 -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -fextend-variable-liveness=none -march=westmere -I/usr/include/p11-kit-1 -DBOOST_ALL_DYN_LINK    -fvisibility-inlines-hidden -Ikmipc/kmipc-2.1.0t-rhel84_64/include -DHAVE_KMIP -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN  -c -o build/dev/test/boost/tablets_test.o test/boost/tablets_test.cc
04:22:50 
 test/boost/tablets_test.cc:1743:62: error: too few arguments to function call, expected at least 3, have 2
04:22:50   1743 |     auto plan = talloc.balance_tablets(stm.get(), stats.get()).get();
04:22:50        |                 ~~~~~~~~~~~~~~~~~~~~~~                       ^
04:22:50  ./service/tablet_allocator.hh:288:28: note: 'balance_tablets' declared here
04:22:50    288 |     future<migration_plan> balance_tablets(locator::token_metadata_ptr, service::topology*, db::system_keyspace*, locator::load_stats_ptr = {}, std::unordered_set<locator::host_id> = {});
04:22:50        |                            ^               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
04:22:50  test/boost/tablets_test.cc:1745:31: error: no viable conversion from '(lambda at test/boost/tablets_test.cc:1745:31)' to 'seastar::noncopyable_function<future<> (token_metadata &)>'
04:22:50   1745 |     stm.mutate_token_metadata([&] (token_metadata& tm) {
04:22:50        |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~
04:22:50   1746 |         return apply_resize_plan(tm, plan);
04:22:50        |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
04:22:50   1747 |     }).get();
04:22:50        |     ~
04:22:50  /jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include/seastar/util/noncopyable_function.hh:203:5: note: candidate constructor not viable: no known conversion from '(lambda at test/boost/tablets_test.cc:1745:31)' to 'const noncopyable_function<seastar::future<> (locator::token_metadata &)> &' for 1st argument
04:22:50    203 |     noncopyable_function(const noncopyable_function&) = delete;
04:22:50        |     ^                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~
04:22:50  /jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include/seastar/util/noncopyable_function.hh:206:5: note: candidate constructor not viable: no known conversion from '(lambda at test/boost/tablets_test.cc:1745:31)' to 'noncopyable_function<seastar::future<> (locator::token_metadata &)> &&' for 1st argument
04:22:50    206 |     noncopyable_function(noncopyable_function&& x) noexcept : _vtable(std::exchange(x._vtable, &_s_empty_vtable)) {
04:22:50        |     ^                    ~~~~~~~~~~~~~~~~~~~~~~~~
04:22:50  /jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include/seastar/util/noncopyable_function.hh:189:5: note: candidate template ignored: constraints not satisfied [with Func = (lambda at test/boost/tablets_test.cc:1745:31)]
04:22:50    189 |     noncopyable_function(Func func) {
04:22:50        |     ^
04:22:50  /jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include/seastar/util/noncopyable_function.hh:188:14: note: because 'std::is_invocable_r_v<seastar::future<>, (lambda at test/boost/tablets_test.cc:1745:31), locator::token_metadata &>' evaluated to false
04:22:50    188 |     requires std::is_invocable_r_v<Ret, Func, Args...>
04:22:50        |              ^
04:22:50  /jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include/seastar/util/noncopyable_function.hh:195:5: note: candidate template ignored: could not match 'seastar::future<> (Object::*)(AllButFirstArg...) noexcept(false)' against '(lambda at test/boost/tablets_test.cc:1745:31)'
04:22:50    195 |     noncopyable_function(Ret (Object::*member)(AllButFirstArg...) noexcept(Noexcept)) : noncopyable_function(std::mem_fn(member)) {}
04:22:50        |     ^
04:22:50  /jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include/seastar/util/noncopyable_function.hh:197:5: note: candidate template ignored: could not match 'seastar::future<> (Object::*)(AllButFirstArg...) const noexcept(false)' against '(lambda at test/boost/tablets_test.cc:1745:31)'
04:22:50    197 |     noncopyable_function(Ret (Object::*member)(AllButFirstArg...) const noexcept(Noexcept)) : noncopyable_function(std::mem_fn(member)) {}
04:22:50        |     ^
04:22:50  ./locator/token_metadata.hh:469:94: note: passing argument to parameter 'func' here
04:22:50    469 |     future<> mutate_token_metadata(seastar::noncopyable_function<future<> (token_metadata&)> func);
04:22:50        |                                                                                              ^
04:22:50  2 errors generated.
```

tgrabiec commented:
In V9:
  - rename per-node flag: "drained" -> "draining"
  - rename feature flag: parallel_decommission -> parallel_tablet_draining
  - adjust diagrams and documentation in topology-over-raft.md
  - add documentation of the new fields added to system tables
  - use row.get_opt() to simplify parsing logic
  - Clarify why we don't cancel paused requests in the topology coordinator scheduler:
    ```patch
    +            // Responsibility for canceling paused requests is on the tablet scheduler.
    +            // Treat them as being able to make progress here.
                 if (topo.paused_requests.contains(req.first)) {
                     continue;
                 }
    ```
  - New commits:
    - af926bb8af doc: topology-over-raft.md: Fix diagram for replacing, tablet_draining is not engaged
    - bd9758df08 test: util: Introduce ensure_group0_leader_on()
    - in python tests:
      - ensure group0 leader is on servers[0]
      - create keyspace using new_test_keyspace()
      - drop redundant tablet-enabling node config
      - add test for canceling of decommission while request is paused
      - Fix typos in patch titles, minor wording changes

tgrabiec commented:
In V9.1:
  - adjust test_tablet_drain_failure_during_decommission to the rename of "drained" to "draining"

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -3196,6 +3198,9 @@ future<service::topology> system_keyspace::load_topology_state(const std::unorde
         sstring cleanup_status = row.get_as<sstring>("cleanup_status");
         utils::UUID request_id = row.get_as<utils::UUID>("request_id");
 
```
nit: is this blank line needed?

tgrabiec replied:
Right, it's not.

gleb-cloudius commented on service/tablet_allocator.cc:
```diff
@@ -1449,6 +1449,18 @@ class load_balancer {
+                // Also, it can fail the removenode request, as failure of this migration is interpreted as
+                // draining failure.
+                // In case of decommission, draining is more important than co-location, so postponing is good.
+                if (nodes.at(dst.host).drained || nodes.at(src.host).drained) {
```
Why do we even consider a node that is been drained as a destination for migration?

tgrabiec commented on db/system_keyspace.cc:
```diff
@@ -3248,6 +3253,9 @@ future<service::topology> system_keyspace::load_topology_state(const std::unorde
 
         if (row.has("topology_request")) {
             auto req = service::topology_request_from_string(row.get_as<sstring>("topology_request"));
+            if (request_paused) {
```
It shouldn't happen, but it's a benign error. request_paused is meaningful only if there is a request. Otherwise, it's ignored.

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -733,6 +748,8 @@ Each node has a clustering row in the table where its `host_id` is the clusterin
 - `node_state`         -  current state of the node (as described earlier)
 - `topology_request`   -  if set contains one of the supported node-specific topology requests
+- `request_paused`     -  set to true when node request should be ignored by the scheduler because it's waiting for the node to be drained of tablet replicas
+- `draining`           -  set to true by `leave` or `remove` requests to indicate to the tablet scheduler that tablets should be migrated away from the node
```
Can we check for removenode/decommission request in  tablet allocator and do not consider such nodes for migration?

tgrabiec commented:
In V10:
 - dropped "draining" column from system.topology, it's derived from node requests.
 - dropped extra space in load_topology_state()
 - dropped cancelation of request if decommissioned node got excluded. not needed anymore.
 - use take_guard() when building topology updates for leave and remove requests
 - include fixes for stats refresh (SBLB), needed for tests to pass:
     -  [topology_coordinator: Refresh load stats after table is created or altered](https://github.com/scylladb/scylladb/pull/27926/changes/266adec472a48bcfb3817d858e2ebe8e3462f995#top)
#27926
    - 0ac1bc00c9 topology_coordinator: Update load stats in case rebuilding with no live replica

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -287,6 +287,7 @@ schema_ptr system_keyspace::topology() {
             .with_column("node_state", utf8_type)
             .with_column("release_version", utf8_type)
             .with_column("topology_request", utf8_type)
+            .with_column("request_paused", boolean_type)
```
Now I wounder if we can figure out paused status as well:

```
if (node.request == leave|remove && has_tablets(node)) {
      node.paused = true;
}
```

tgrabiec replied:
That's not a strong point, since we already do a lot of work on each stage transition (load balancing).

tgrabiec replied:
v11 implements this suggestion

tgrabiec commented on service/topology_coordinator.cc:
```diff
@@ -1185,6 +1192,22 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+        auto nodes = get_barrier_nodes();
+        co_await seastar::parallel_for_each(std::move(nodes), [this] (raft::server_id id) {
+            // cmd_index is intentionally 0 so that it doesn't disrupt regular command sequencing.
+            // The command will be rejected by the handler but that's ok, we only want the group0 read barrier.
```
Changed in v11

gleb-cloudius commented on service/tablet_allocator.cc:
```diff
@@ -592,6 +592,13 @@ class load_balancer {
         std::optional<disk_usage> dusage; // Invariant: bool(dusage) || drained.
         bool drained = false;
+
+        // Engaged if and only if drained == true.
```
nit: can those two be folded then?

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -3174,29 +3197,51 @@ class topology_coordinator : public endpoint_lifecycle_subscriber
+                            co_await update_topology_state(take_guard(std::move(node)), {builder.build(), rtbuilder.build()},
+                                                           "start decommission");
+                        } else {
+                            builder.set_transition_state(topology::transition_state::tablet_migration)
```
Why would we have tablets here though? Tablets should have been drained before we get here, no? And if they are not we should move to `topology::transition_state::tablet_draining` which is now deprecated. Why `tablet_migration` would drain them?

gleb-cloudius commented on service/topology_state_machine.cc:
```diff
@@ -318,6 +321,56 @@ void topology_state_machine::generate_cancel_request_update(utils::chunked_vecto
+            }
+
+            if (req == topology_request::join || req == topology_request::replace) {
+                // Only topology coordinator can abort join requests. It sends reject RPC.
```
> If we make it go through the coordinator, there would be no way to cancel something once it takes over the state machine. We may want to interrupt vnode streaming with this, for example.

The code does not allow to cancel requests after they started.

> 
> I think that the reject RPC should be replaced with a reaction to group0 state (e.g. scheduling a new "reject" request). This solves not only cancelation problem, but also coordinator failover. Currently, if the coordinator failovers after canceling the request but before RPC succeeds, reject RPC will not be resent.

Reject rpc delivery is not mandatory. It makes for better UI, but with a change that aborts a node if it is banned the joining node will notice anyway.

> 
> Any action which is based on transient state is probably wrong, because of failover. And if all is based on group0 state, we can do it from non-coordinator.

Doing things outside of the coordinator is error prone. You can modify some state but not other. It is unavoidable since placing a request is also state modification, but at least request part was designed to be modifiable by non coordinator.

Anyway I am not insisting on changing it here.

tgrabiec commented:
The failures about expected error message are introduced by the PR, I need to adjust test expectations.

paszkow commented on service/topology_coordinator.cc:
```diff
@@ -1633,6 +1637,19 @@ class topology_coordinator : public endpoint_lifecycle_subscriber
+                            auto req = _topo_sm._topology.get_request(raft::server_id(maybe_leaving->host.uuid()));
+                            // We only cancel remove request on critical utilization to avoid failing remove operation
+                            // once migrations started before the node was marked as excluded are unblocked and roll back.
+                            if (req == topology_request::leave || critical_disk_utilization) {
```
I believe it should be `&&`. Otherwise, you will always cancel the remove request.

tgrabiec replied:
I want to cancel `remove` request when there is critical_disk_utilization, because it won't be able to progress. The failure is not transient. I don't want to cancel remove if the failure could be due to node being down.

patjed41 commented on test/cluster/test_tablets_parallel_decommission.py:
```diff
@@ -0,0 +1,539 @@
+        {"dc": "dc1", "rack": "rack1"},
+        {"dc": "dc1", "rack": "rack2"},
+    ])
+    await ensure_group0_leader_on(manager, servers[0])
```
Why not do this at the end of `servers_add`?

patjed41 reviewed (commented):
Only some comments related to tests.

I don't see any critical problem, so I can merge this PR as is, but I'll wait for responses. AFAIK, the plan is to branch 2026.1 in two days. If you send a new version, I can ignore any unrelated CI failures to make sure we merge this PR in time.
